hypershift-operator/controllers/scheduler/autoscaler.go

```diff
@@ -653,3 +728,121 @@ func podNames(pods []corev1.Pod) []string {
+}
+
+func validateNonRequestServingMachineSets(machineSets []machinev1beta1.MachineSet) error {
+	if len(machineSets) != 3 {
+		return fmt.Errorf("expected 3 non request serving machinesets, found %d", len(machineSets))
+	}
+
+	// Ensure that we have consistent min/max size across machinesets
+	minSizes := sets.New[int]()
+	maxSizes := sets.New[int]()
+	for _, ms := range machineSets {
+		minSize, err := strconv.Atoi(ms.Annotations[minSizeMachineSetAnnotation])
+		if err != nil {
+			return fmt.Errorf("failed to parse min size annotation (%q): %w", ms.Annotations[minSizeMachineSetAnnotation], err)
+		}
+		minSizes.Insert(minSize)
+		maxSize, err := strconv.Atoi(ms.Annotations[maxSizeMachineSetAnnotation])
+		if err != nil {
+			return fmt.Errorf("failed to parse max size annotation (%q): %w", ms.Annotations[maxSizeMachineSetAnnotation], err)
+		}
+		maxSizes.Insert(maxSize)
+	}
+	if minSizes.Len() != 1 || maxSizes.Len() != 1 {
+		return fmt.Errorf("inconsistent min/max sizes across non request serving machinesets")
+	}
+
+	minSize := minSizes.UnsortedList()[0]
+	maxSize := maxSizes.UnsortedList()[0]
+
+	if maxSize < minSize || maxSize < 1 {
+		return fmt.Errorf("invalid min/max sizes, minSize: %d, maxSize: %d", minSize, maxSize)
+	}
+	return nil
+}
+
+type machineSetReplicas struct {
+	machineSet machinev1beta1.MachineSet
+	replicas   int32
+}
+
+func nonRequestServingMachineSetsToScale(ctx context.Context, config *schedulingv1alpha1.ClusterSizingConfiguration, hostedClusters []hyperv1.HostedCluster, machineSets []machinev1beta1.MachineSet) []machineSetReplicas {
+	log := ctrl.LoggerFrom(ctx)
+	hcCountBySize := make(map[string]int)
+	notLabeled := 0
+	for _, hc := range hostedClusters {
+		sizeLabelValue := hc.Labels[hyperv1.HostedClusterSizeLabel]
+		if sizeLabelValue == "" {
+			notLabeled++
+			continue
+		}
+		hcCountBySize[sizeLabelValue]++
+	}
+	if notLabeled > 0 {
+		log.Info("WARNING: Hosted clusters without size label. Using the smallest size for them.", "count", notLabeled)
+		for _, sizeConfig := range config.Spec.Sizes {
+			if sizeConfig.Criteria.From == 0 {
+				hcCountBySize[sizeConfig.Name] += notLabeled
+				break
+			}
+		}
+	}
+
+	// calculate the number of non request serving nodes required per zone
+	nodesNeededQty := resource.MustParse("0")
+	for size, count := range hcCountBySize {
+		sizeConfigFound := false
+		for _, sizeConfig := range config.Spec.Sizes {
+			if sizeConfig.Name == size {
+				sizeConfigFound = true
+				nodesForSize := *sizeConfig.Management.NonRequestServingNodesPerZone
+				nodesForSize.Mul(int64(count))
+				nodesNeededQty.Add(nodesForSize)
+				break
+			}
+		}
+		if !sizeConfigFound {
+			log.Info("WARNING: No size configuration found for hosted cluster size", "size", size)
+		}
+	}
+	if config.Spec.NonRequestServingNodesBufferPerZone != nil {
+		nodesNeededQty.Add(*config.Spec.NonRequestServingNodesBufferPerZone)
+	}
+	nodesNeeded := int32(math.Ceil(nodesNeededQty.AsApproximateFloat64()))
```

Comment:
Why did we need the resource quantity logic on the API if we're going to squash into an int32 anyway? I thought the utility there was e.g. `128Mi`

Reply:
Because we want to allow decimal values

Reply:
I see